### PR TITLE
 [FEATURE] Feature filter algorithm for processing models

### PR DESCRIPTION
**Pull request by m-kuhn**
## Description
Adds a new "feature filter" algorithm. This algorithm takes a list of expressions and
creates an output for each of them. Matching features from the input layer are sent
to the outputs.

This allows for an algorithm to only handle a subset of incoming features or to send
different features to different outputs.

The algorithm is only available in the modeler.

image

## TODO

- [x] Make it work
- [x] Unit Tests
- [x] Allow outputs to be directly written to new layers
- [x] Have an expression builder widget to define the expression

### CODE REVIEW

**Comment by nyalldawson** [comment]
> Allow outputs to be directly written to new layers

What's this to-do mean?

> Unit tests

Heads up - some C++ native modeler algs have direct C++ unit tests. This is probably an easier approach to use here also and avoid the complexity of trying to fit the tests into the normal algorithm test framework

**Comment by m-kuhn** [comment]
> >  Allow outputs to be directly written to new layers
> What's this to-do mean?

So far outputs can only be sent to other algorithms. If nothing is connected, anything sent there is gone for good.

> > Unit tests
> Heads up - some C++ native modeler algs have direct C++ unit tests. This is probably an easier approach to use here also and avoid the complexity of trying to fit the tests into the normal algorithm test framework

Is it not possible to call a model from the test framework? That would make it quite easy to test.

**Comment by m-kuhn** [comment]
The hard part is the expression builder. Because currently the config widget is created within the analysis library, qgis_gui (and therefore the nice expression widgets) are not available. I am a bit worried that this will considerably complicate things.

**Comment by nyalldawson** [comment]
> Is it not possible to call a model from the test framework? That would make it quite easy to test

You can, but adding specific granular tests would be good here (e.g. checking that outputs are correctly created after init with different configurations). That'd be a pain to do using standard algorithm tests.

File: src/analysis/processing/qgsalgorithmfilter.cpp [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+}
+
+QStringList QgsFilterAlgorithm::tags() const
+{
+  return QObject::tr( "filter,proxy,redirect,route" ).split( ',' );
+}
+
+QString QgsFilterAlgorithm::group() const
+{
+  return QObject::tr( "Vector geometry" );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmfilter.cpp]: This should be Vector table, not geometry

File: src/analysis/processing/qgsalgorithmfilter.cpp [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+}
+
+
+QVariantMap QgsFilterAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  std::unique_ptr< QgsFeatureSource > source( parameterAsSource( parameters, QStringLiteral( "INPUT" ), context ) );
+  if ( !source )
+    return QVariantMap();
+
+  QgsExpressionContext expressionContext = context.expressionContext();
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmfilter.cpp]: Use createExpressionContext instead

File: src/analysis/processing/qgsalgorithmfilter.cpp [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+
+    feedback->setProgress( current * step );
+    current++;
+  }
+
+  QVariantMap outputs;
+  for ( const Output *output : qgis::as_const( mOutputs ) )
+  {
+    outputs.insert( output->name, output->destinationIdentifier );
+  }
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmfilter.cpp]: I think you should be deleting the sinks here too

Comment (m-kuhn) [reply src/analysis/processing/qgsalgorithmfilter.cpp]: They are deleted in the destructor, but that can also be moved here.

Comment (nyalldawson) [reply src/analysis/processing/qgsalgorithmfilter.cpp]: I think it needs to be moved - some sinks do finalization code which probably should be done be within the processAlgorithm step

Comment (m-kuhn) [reply src/analysis/processing/qgsalgorithmfilter.cpp]: also done

File: src/core/processing/models/qgsprocessingmodelalgorithm.cpp [hunk 4]
```diff
@@ -89,9 +89,6 @@ QVariantMap QgsProcessingModelAlgorithm::parametersForChildAlgorithm( const QgsP
   QVariantMap childParams;
   Q_FOREACH ( const QgsProcessingParameterDefinition *def, child.algorithm()->parameterDefinitions() )
   {
-    if ( def->flags() & QgsProcessingParameterDefinition::FlagHidden )
-      continue;
```
Comment (m-kuhn) [inline src/core/processing/models/qgsprocessingmodelalgorithm.cpp]: @nyalldawson this fixes a crash for me as soon as I set the outputs to hidden (so I don't see them as parameters in the algorithm configuration widget). I am 0% sure this is the right approach to deal with this :D

**Comment by nyalldawson** [comment]
Looking very nice!

The one thing I don't like here is that QgsProcessingAlgorithmConfigurationWidget and QgsProcessingAlgorithmConfigurationWidgetFactory are in core, when they really belong in gui.

What do you think about moving these to gui, and then creating a separate registry for the gui components of processing? That's how I handled this recently with the layouts engine - there's QgsLayoutItemRegistry in core (and an instance attached to QgsApplication) and QgsLayoutItemGuiRegistry in gui (with an instance attached to QgsGui).

This means there could be a total clean break between the processing gui and core components.

**Comment by m-kuhn** [comment]
Sounds good to me, I was also thinking about that.

File: src/core/processing/models/qgsprocessingmodelalgorithm.cpp [hunk 5]
```diff
@@ -769,6 +769,9 @@ void QgsProcessingModelAlgorithm::updateDestinationParameters()
         continue;
 
       std::unique_ptr< QgsProcessingParameterDefinition > param( source->clone() );
+      // Even if an output was hidden in a child algorithm, we want to show it here for the final
+      // outputs.
+      param->setFlags( param->flags() & ~QgsProcessingParameterDefinition::FlagHidden );
```
Comment (m-kuhn) [inline src/core/processing/models/qgsprocessingmodelalgorithm.cpp]: Another very hacky-feeling thing to do. Does someone know what hidden parameters are for?

File: src/analysis/processing/qgsalgorithmfilter.cpp [hunk 6]
```diff
@@ -92,6 +96,8 @@ QVariantMap QgsFilterAlgorithm::processAlgorithm( const QVariantMap &parameters,
   for ( Output *output : qgis::as_const( mOutputs ) )
   {
     output->sink.reset( parameterAsSink( parameters, output->name, context, output->destinationIdentifier, source->fields(), source->wkbType(), source->sourceCrs() ) );
+    if ( !output->sink )
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmfilter.cpp]: I think this should throw a QgsProcessingException instead - it's a fatal error, we don't want to silently ignore it and continue with the model execution.

**Comment by nyalldawson** [comment]
:+1:

File: src/analysis/processing/qgsalgorithmfilter.cpp (hunk 6 above)
Comment (m-kuhn) [reply src/analysis/processing/qgsalgorithmfilter.cpp]: That was just copied from the buffer algorithm, I also found that a bit strange.

Comment (nyalldawson) [reply src/analysis/processing/qgsalgorithmfilter.cpp]: I'll fix that

Comment (m-kuhn) [reply src/analysis/processing/qgsalgorithmfilter.cpp]: Cool, thanks. The one for here is in the pipeline too

Comment (nyalldawson) [reply src/analysis/processing/qgsalgorithmfilter.cpp]: https://github.com/qgis/QGIS/commit/015665688262f95e148972d699fc2270d622a75c

File: src/analysis/processing/qgsalgorithmfilter.cpp [hunk 7]
```diff
@@ -88,6 +88,7 @@ void QgsFilterAlgorithm::initAlgorithm( const QVariantMap &configuration )
 
 QVariantMap QgsFilterAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
 {
+  context.setInvalidGeometryCheck( QgsFeatureRequest::GeometryNoCheck );
```
Comment (m-kuhn) [inline src/analysis/processing/qgsalgorithmfilter.cpp]: @nyalldawson is this the correct place to disable this check? I don't think it's used anywhere so far...

**Comment by m-kuhn** [comment]
Any final comments on this?

**Comment by nyalldawson** [comment]
Looks fantastic to me! Could you add tests covering QgsProcessingGuiRegistry and then I'm a big +1 to merge.

**Pull request merged**